### PR TITLE
Add adversarial caunet

### PR DESCRIPTION
**Pull request by mosem**

### CODE REVIEW

File: conf/experiment/caunet_1.yaml [hunk 1]
```diff
@@ -1,4 +1,4 @@
-experiment_name: demucs_1
+experiment_name: caunet_1
```
Comment (Or-Tal) [inline conf/experiment/caunet_1.yaml:1]: see prev pull requests

File: conf/experiment/demucs_1.yaml [hunk 2]
```diff
@@ -27,4 +27,6 @@ demucs:
   resample: 1
   growth: 2
   rescale: 0.1
-  scale_factor: ${experiment.scale_factor}
\ No newline at end of file
+  scale_factor: ${experiment.scale_factor}
+
+adversarial: False
```
Comment (Or-Tal) [inline conf/experiment/demucs_1.yaml:32]: see prev pull requests

File: conf/experiment/example_file.yaml [hunk 3]
```diff
@@ -80,4 +80,6 @@ discriminator:
   n_layers: 4
   num_D: 3
   downsampling_factor: 4
-  ndf: 16
```
Comment (Or-Tal) [inline conf/experiment/example_file.yaml:83]: what are theses flags? why are they in the general example file? these are configuration specific.

File: conf/experiment/example_file.yaml [hunk 4]
```diff
@@ -80,4 +80,6 @@ discriminator:
...
\ No newline at end of file
+  ndf: 16
```
Comment (Or-Tal) [inline conf/experiment/example_file.yaml:83]: dido

File: conf/experiment/example_file.yaml [hunk 5]
```diff
@@ -80,4 +80,6 @@ discriminator:
...
+
+adversarial: False
```
Comment (Or-Tal) [inline conf/experiment/example_file.yaml:85]: dido

File: denoiser/batch_solvers/adversarial_bs.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+class AdversarialBS(GeneratorBS):
+
+    def __init__(self, args, generator, discriminator):
+        super().__init__(args, generator)
+        if torch.cuda.is_available():
+            discriminator.cuda()
+        self._models.update({'discriminator': discriminator})
+
+        disc_optimizer = torch.optim.Adam(discriminator.parameters(), lr=self.args.lr, betas=(0.9, self.args.beta2))
+        self._optimizers.update(({'discriminator_optimizer': disc_optimizer}))
```
Comment (Or-Tal) [inline denoiser/batch_solvers/adversarial_bs.py:16]: use magic number instead of explicit strings

File: denoiser/batch_solvers/adversarial_bs.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+        self._losses_names += ['discriminator']
```
Comment (Or-Tal) [inline denoiser/batch_solvers/adversarial_bs.py:17]: dido

File: denoiser/batch_solvers/adversarial_bs.py [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+
+    def run(self, data, cross_valid=False):
+        noisy, clean = data
+
+        generator = self._models['generator']
```
Comment (Or-Tal) [inline denoiser/batch_solvers/adversarial_bs.py:23]: magic num instead of explicit string

File: denoiser/batch_solvers/adversarial_bs.py [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+
+        losses = {self._losses_names[0]: total_loss_generator.item(), self._losses_names[1]: loss_discriminator.item()}
+
+        if not cross_valid:
+            self._optimize(total_loss_generator, loss_discriminator)
+
+        return losses
+
+    def _optimize(self, total_loss_G, loss_D):
+        generator_optimizer = self._optimizers['generator_optimizer']
```
Comment (Or-Tal) [inline denoiser/batch_solvers/adversarial_bs.py:43]: magic num instead of explicit string

File: denoiser/batch_solvers/batch_solver_factory.py [hunk 10]
```diff
@@ -1,20 +1,43 @@
...
     def get_bs(args):
-        if args.experiment.model == "demucs":
-            generator = Demucs(**args.experiment.demucs)
-            return GeneratorBS(args, generator)
-        if args.experiment.model == "caunet":
-            generator = Caunet(**args.experiment.caunet)
-            return GeneratorBS(args, generator)
-        elif args.experiment.model == "demucs_hifi":
-            return DemucsHifiBS(args)
+        if args.experiment.adversarial:
```
Comment (Or-Tal) [inline denoiser/batch_solvers/batch_solver_factory.py:14]: see prev comment about this redundancy

File: denoiser/models/modules.py [hunk 11]
```diff
@@ -1,6 +1,6 @@
 import math
 import torch
-from torch import nn, nn as nn
```
Comment (Or-Tal) [inline denoiser/models/modules.py:3]: see prev comment about long file

File: launch_valentini.sh [hunk 12]
```diff
@@ -8,7 +8,7 @@
```
Comment (Or-Tal) [inline launch_valentini.sh:8]: see prev comment

**Review by Or-Tal — CHANGES_REQUESTED** [review]
Hi, using explicit strings is asking for troubles, I think you should replace all strings with magic numbers that are defined in a single point of truth.
Please read the rest of my notes

File: conf/experiment/example_file.yaml (hunk 3 above)
Comment (mosem) [reply conf/experiment/example_file.yaml]: I don't see any difference between discriminators and other models, they should be here just like any other model.

**Review by mosem — COMMENTED** [review]
looks done. merging...

**Pull request merged**